subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c

```diff
@@ -63,23 +63,67 @@ struct shi_it8xxx2_cfg {
 	const struct gpio_dt_spec cs;
 };
 
+enum shi_ite_pm_policy_state_flag {
+	SHI_ITE_PM_POLICY_FLAG,
+	SHI_ITE_PM_POLICY_FLAG_COUNT,
+};
+
 struct shi_it8xxx2_data {
 	/* Peripheral data */
 	struct ec_host_cmd_rx_ctx *rx_ctx;
 	struct ec_host_cmd_tx_buf *tx;
 	struct gpio_callback cs_cb;
+	struct k_work_delayable cs_off_put;
 	/* Current state */
 	enum shi_state_machine shi_state;
 	/* Buffers */
 	uint8_t in_msg[SPI_RX_MAX_FIFO_SIZE] __aligned(4);
 	uint8_t out_msg[SPI_TX_MAX_FIFO_SIZE] __aligned(4);
+	ATOMIC_DEFINE(pm_policy_state_flag, SHI_ITE_PM_POLICY_FLAG_COUNT);
 };
 
 struct ec_host_cmd_shi_ite_ctx {
 	/* SHI device instance */
 	const struct device *dev;
 };
 
+static void shi_ite_pm_policy_state_lock_get(struct shi_it8xxx2_data *data,
+					     enum shi_ite_pm_policy_state_flag flag)
+{
+	if (atomic_test_and_set_bit(data->pm_policy_state_flag, flag) == 0) {
+		pm_policy_state_lock_get(PM_STATE_STANDBY, PM_ALL_SUBSTATES);
+		k_work_reschedule(&data->cs_off_put, K_SECONDS(3));
+	}
+}
+
+static void shi_ite_pm_policy_state_lock_put(struct shi_it8xxx2_data *data,
+					     enum shi_ite_pm_policy_state_flag flag)
+{
+	if (atomic_test_and_clear_bit(data->pm_policy_state_flag, flag) == 1) {
+		pm_policy_state_lock_put(PM_STATE_STANDBY, PM_ALL_SUBSTATES);
+		k_work_cancel_delayable(&data->cs_off_put);
+	}
+}
+
+/*
+ * When AP is In S0 state, AP assert CS of SPI for a new command transaction.
+ * Under the condition, SoC will not enter deep sleep until AP de-assert the CS.
+ * But If AP is powered off (G3 state) CS will go low (assert CS).
+ * This assertion will prevent SoC from entering deep sleep but the assertion
+ * is not for starting a new command transaction.
+ * This handler is used to resolve the situation.
+ */
+static void cs_off_put_handler(struct k_work *item)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(item);
+	struct shi_it8xxx2_data *data = CONTAINER_OF(dwork, struct shi_it8xxx2_data, cs_off_put);
+	int key;
+
+	key = irq_lock();
```

Comment:
Why do we need irq_lock here?

Reply:
The above pm policy get/put is called in ISR, and this put is called in workqueue thread.
I want to make sure there is no potential race condition between them.

Reply:
I don't it is necessary. Please correct me if I'm wrong. The handler can be called only if CS is asserted (de-assertion cancels k_work). So we stop irq only low->high, which will do the same anyway. So I think it is not needed, especially with the atomic-test guard.

Reply:
Removed the irq_lock.

Reply:
Thanks, I hope I haven't missed anything.